# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
os.getcwd()

'C:\\Users\\craig\\Documents\\Python Scripts\\Notebooks'

In [3]:
path="C:\\Users\\craig\\Documents\\Python Scripts\\Notebooks\\Step 2\\data"
os.chdir(path) 

In [9]:
df=pd.read_csv('step3_output.csv', index_col=0)

In [10]:
df.head()

Name       state  summit_elev  vertical_drop  \
2               Hilltop Ski Area      Alaska         2090            294   
4            Sunrise Park Resort     Arizona        11100           1800   
5  Yosemite Ski & Snowboard Area  California         7800            600   
8         Boreal Mountain Resort  California         7700            500   
9                    Dodge Ridge  California         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple    ...     \
2      0        0.0          0          0     0       1    ...      
4      0        0.0          0          1     2       3    ...      
5      0        0.0          0          0     0       1    ...      
8      0        0.0          0          1     1       3    ...      
9      0        0.0          0          0     1       2    ...      

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
2               30.0            30.0        150.000000       36.0   
4              800.0            80.0        115.000000       49.0   
5               88.0             0.0        110.000000       84.0   
8              380.0           200.0        150.000000       54.0   
9              862.0             0.0        151.578947       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
2             69.0          30.0     34.000000              152.0   
4            250.0          74.0     78.000000              104.0   
5            300.0          47.0     47.000000              107.0   
8            400.0          49.0     89.538462              150.0   
9            350.0          78.0     78.000000              140.0   

   NightSkiing_ac  clusters  
2            30.0         2  
4            80.0         1  
5             0.0         1  
8           200.0         1  
9             0.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [26]:
dfo=df['state'] # select object type columns
#dfo2=df.select_dtypes(include=['object'])

In [34]:
df = df.drop('state', axis=1)

In [35]:
df = pd.concat([df, pd.get_dummies(dfo)], axis=1)

In [93]:
print(df.columns)

Index(['Name', 'summit_elev', 'vertical_drop', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters', 'Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'South Dakota', 'Tennessee', 'Utah', 'Vermont',
       'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming',
       'Alaska', 'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho',
       'Illinois', 'Indiana', 'Iowa', 'Maine', 'Maryland',

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [94]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend','trams', 'fastEight',
       'fastSixes'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [95]:
X_scaled.std(axis = 0)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1.])

In [96]:
print(X_scaled)

[[-0.4851081  -0.93139617 -0.40485066 ... -0.13284223 -0.26057865
  -0.17251639]
 [ 2.33098716  1.64953271  1.30342163 ... -0.13284223 -0.26057865
  -0.17251639]
 [ 1.29956492 -0.40698433 -0.40485066 ... -0.13284223 -0.26057865
  -0.17251639]
 ...
 [ 1.30206534  1.25708071 -0.40485066 ... -0.13284223 -0.26057865
   5.7965507 ]
 [ 1.88184966  0.26138371 -0.40485066 ... -0.13284223 -0.26057865
   5.7965507 ]
 [ 1.83090365  0.44989777 -0.40485066 ... -0.13284223 -0.26057865
   5.7965507 ]]


## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [97]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled,y, test_size=0.25, random_state=1)

In [117]:
#np.delete(arr, 1, 0)

np.set_printoptions(threshold=np.nan)
y_test2 = np.delete(y_test,13)
print(y_test2)

[80. 64. 60. 59. 49. 45. 68. 89. 45. 59. 79. 78. 67. 30. 49. 75. 40. 39.
 40. 49. 70. 60. 50. 32. 45. 47. 55. 84. 49. 47. 92. 79. 75. 69. 49. 20.
 70. 43. 35. 34. 42. 65. 49.]


In [119]:
#np.delete(arr, 1, 0)

np.set_printoptions(threshold=np.nan)
y_test3 = np.delete(y_test2,20)
print(y_test3)

[80. 64. 60. 59. 49. 45. 68. 89. 45. 59. 79. 78. 67. 30. 49. 75. 40. 39.
 40. 49. 60. 50. 32. 45. 47. 55. 84. 49. 47. 92. 79. 75. 69. 49. 20. 70.
 43. 35. 34. 42. 65. 49.]


In [126]:
len(y_test3)

42

In [132]:
np.set_printoptions(threshold=np.nan)
X_test2 = np.delete(X_test,13,0)
print(X_test2)

[[ 2.63572555e+00  1.52100040e+00 -4.04850659e-01  6.51758391e-01
   6.48294699e-01  1.87534342e-01 -1.88271306e-01  3.70406385e-01
   3.07442101e+00 -5.46183960e-01  2.74109980e+00  1.50864575e+00
   2.71916638e+00  9.09098285e-02  9.57683227e-01  7.62939154e-01
   1.79907505e+00  8.75443517e-01 -9.35740239e-01 -5.60082153e-01
  -7.62492852e-02 -7.62492852e-02 -2.73009453e-01 -1.53846154e-01
  -1.72516390e-01 -1.89547207e-01 -1.32842233e-01 -1.08147614e-01
  -1.32842233e-01 -1.72516390e-01 -7.62492852e-02 -1.53846154e-01
  -3.08118011e-01 -2.20192753e-01 -1.08147614e-01 -1.32842233e-01
  -7.62492852e-02 -2.60578653e-01 -7.62492852e-02  5.27573060e+00
  -3.61550763e-01 -1.53846154e-01 -1.53846154e-01 -1.32842233e-01
  -2.60578653e-01 -7.62492852e-02 -7.62492852e-02 -1.32842233e-01
  -2.20192753e-01 -1.08147614e-01 -1.32842233e-01 -1.32842233e-01
  -2.60578653e-01 -1.72516390e-01 -7.62492852e-02 -7.62492852e-02
  -2.73009453e-01 -1.53846154e-01 -1.72516390e-01 -1.89547207e-01
  -1.32842

In [133]:
np.set_printoptions(threshold=np.nan)
X_test3 = np.delete(X_test2,20,0)
print(X_test3)

[[ 2.63572555e+00  1.52100040e+00 -4.04850659e-01  6.51758391e-01
   6.48294699e-01  1.87534342e-01 -1.88271306e-01  3.70406385e-01
   3.07442101e+00 -5.46183960e-01  2.74109980e+00  1.50864575e+00
   2.71916638e+00  9.09098285e-02  9.57683227e-01  7.62939154e-01
   1.79907505e+00  8.75443517e-01 -9.35740239e-01 -5.60082153e-01
  -7.62492852e-02 -7.62492852e-02 -2.73009453e-01 -1.53846154e-01
  -1.72516390e-01 -1.89547207e-01 -1.32842233e-01 -1.08147614e-01
  -1.32842233e-01 -1.72516390e-01 -7.62492852e-02 -1.53846154e-01
  -3.08118011e-01 -2.20192753e-01 -1.08147614e-01 -1.32842233e-01
  -7.62492852e-02 -2.60578653e-01 -7.62492852e-02  5.27573060e+00
  -3.61550763e-01 -1.53846154e-01 -1.53846154e-01 -1.32842233e-01
  -2.60578653e-01 -7.62492852e-02 -7.62492852e-02 -1.32842233e-01
  -2.20192753e-01 -1.08147614e-01 -1.32842233e-01 -1.32842233e-01
  -2.60578653e-01 -1.72516390e-01 -7.62492852e-02 -7.62492852e-02
  -2.73009453e-01 -1.53846154e-01 -1.72516390e-01 -1.89547207e-01
  -1.32842

In [134]:
len(X_test3)

42

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [157]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [100]:
print('Intercept: \n', model.intercept_)
print('Coefficients: \n', model.coef_)

Intercept: 
 73032336319255.56
Coefficients: 
 [ 5.39557675e-01  1.42402932e+00  5.53431399e+13  6.62392260e+13
  1.16327235e+14  1.39873428e+14  1.36423288e+14 -2.61137710e+14
  7.10028199e-01  1.08984375e+00  2.34375000e-02 -3.28125000e-01
 -6.01562500e-01  1.11718750e+00 -1.09375000e+00  1.93750000e+00
  7.70312500e+00 -5.46875000e-01  1.60937500e+00 -7.26562500e-01
 -5.83110352e+14 -3.37224746e+13 -2.29323548e+14 -1.43199746e+14
 -9.75390572e+13  3.64997894e+14 -5.87745331e+13 -2.73464699e+14
 -9.68513671e+13 -2.64605214e+14  2.04648017e+14 -2.64215591e+14
 -3.52525132e+14 -2.54803037e+14 -5.22759716e+13 -2.29351232e+14
  2.18607804e+14 -1.45592131e+14  2.55399951e+14 -2.37858894e+14
 -5.76416147e+13 -2.78807537e+13  1.48910758e+14 -1.43449575e+14
 -9.56343086e+13  1.46958599e+14  1.45718175e+13  1.37260502e+14
 -2.87877897e+14 -7.94747555e+10 -2.02680818e+13 -4.17307807e+13
 -2.79283264e+14 -3.37075398e+14  5.20654472e+14 -2.87334044e+13
  2.00026142e+13  1.93821266e+13 -4.0483079

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [135]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test

y_pred = model.predict(X_test3)

In [136]:
print(y_pred)

[71.0625   58.75     63.625    64.4375   50.4375   48.375    58.4375
 78.6875   49.5625   55.5      87.0625   86.0625   57.4375   34.40625
 62.9375   49.9375   47.9375   34.0625   35.1875   57.875    62.1875
 48.3125   47.5625   48.25     51.875    55.375    77.4375   42.125
 46.21875  73.234375 89.5625   75.8125   52.9375   56.0625   34.4375
 75.9375   43.3125   34.5625   31.6875   40.8125   58.4375   49.5625  ]


In [108]:
print(X_test)

[[ 2.63572555  1.5210004  -0.40485066 ... -0.13284223 -0.26057865
  -0.17251639]
 [ 0.11186659 -0.80971891 -0.40485066 ... -0.13284223 -0.26057865
  -0.17251639]
 [-0.97425229 -0.70689306 -0.40485066 ... -0.13284223 -0.26057865
  -0.17251639]
 ...
 [-0.76327956 -1.0496459  -0.40485066 ... -0.13284223 -0.26057865
  -0.17251639]
 [ 0.33065312 -0.20133262 -0.40485066 ... -0.13284223 -0.26057865
  -0.17251639]
 [-0.43509975  0.53558598 -0.40485066 ... -0.13284223 -0.26057865
  -0.17251639]]


## Review Model Outcomes — Iterate over additional models as needed

In [0]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [138]:
vs = explained_variance_score(y_test3, y_pred)

In [139]:
print(vs)

0.7495868442012101


In [142]:
ae = mean_absolute_error(y_test3, y_pred)

In [143]:
print(ae)

6.638764880952381


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [144]:
print(model.intercept_)

73032336319255.56


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [145]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coefdf= pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


In [151]:
coefdf.sort_values(by='Coefficient', ascending=False)

Coefficient
Alaska             5.831104e+14
Alaska             5.206545e+14
Idaho              5.157429e+14
Idaho              3.649979e+14
Michigan           3.525251e+14
Wyoming            3.370754e+14
New Jersey         3.178558e+14
Vermont            2.878779e+14
Wisconsin          2.792833e+14
Indiana            2.734647e+14
Ohio               2.727284e+14
Maine              2.646052e+14
Massachusetts      2.642156e+14
total_chairs       2.611377e+14
New Jersey         2.554000e+14
Minnesota          2.548030e+14
Utah               2.448065e+14
New Mexico         2.378589e+14
Montana            2.293512e+14
California         2.293235e+14
Nevada             2.186078e+14
South Dakota       2.094145e+14
New York           2.057912e+14
Maryland           2.048213e+14
Nevada             2.048213e+14
Maryland           2.046480e+14
Wyoming            1.990533e+14
Indiana            1.853959e+14
Ohio               1.489108e+14
South Dakota       1.469586e+14
...                         ...
fastQuads          5.534314e+13
Missouri           5.227597e+13
Illinois           4.877145e+13
West Virginia      4.173078e+13
Connecticut        4.048308e+13
Oregon             3.590359e+13
Missouri           3.579284e+13
Arizona            3.372247e+13
Arizona            2.873340e+13
North Carolina     2.788075e+13
Washington         2.026808e+13
California         2.000261e+13
Colorado           1.938213e+13
Tennessee          1.457182e+13
Iowa               1.069462e+13
Virginia           7.947476e+10
AdultWeekday       7.703125e+00
averageSnowfall    1.937500e+00
NightSkiing_ac     1.609375e+00
vertical_drop      1.424029e+00
daysOpenLastYear   1.117188e+00
yearsOpen          1.093750e+00
TerrainParks       1.089844e+00
clusters           7.265625e-01
Runs               7.100282e-01
Snow Making_ac     6.015625e-01
projectedDaysOpen  5.468750e-01
summit_elev        5.395577e-01
SkiableTerrain_ac  3.281250e-01
LongestRun_mi      2.343750e-02

[88 rows x 1 columns]

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [152]:
df2=pd.read_csv('step3_output.csv', index_col=0)

In [153]:
df2 = df2.drop('state', axis=1)

In [154]:
X2 = df2.drop(['Name','AdultWeekend'], axis=1)
y2 = df2.AdultWeekend
scaler2 = preprocessing.StandardScaler().fit(X2)
X_scaled2=scaler2.transform(X2) 

In [155]:
y2 = y2.ravel()
X2_train, X2_test, y2_train, y2_test = train_test_split(X_scaled2,y2, test_size=0.25, random_state=1)

In [158]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X2_train,y2_train)

In [159]:
y2_pred = model2.predict(X2_test)

In [160]:
print(y2_pred)

[72.97651466 54.96395234 60.27769582 67.60393626 55.11207893 47.64116465
 58.67333255 73.2366414  50.62782976 57.13268225 80.68428235 79.7283418
 59.79974626 75.84669396 32.06807937 58.80854316 49.32559319 50.43867598
 45.30877085 43.1294821  60.36159357 64.57513327 63.87637963 53.51799745
 42.05961651 40.7645932  53.9504888  56.18084144 74.96102412 39.0461085
 51.53275712 75.2057254  93.07355442 70.38025625 56.29512983 50.40159575
 26.94757055 68.69588874 53.05294471 41.38134183 34.87542758 44.62312539
 53.88842504 41.37090572]


In [161]:
print(model2.intercept_)

57.16177829831384


In [162]:
vs2 = explained_variance_score(y2_test, y2_pred)

In [163]:
print(vs2)

0.7542459520255993


In [164]:
print(mean_absolute_error(y2_test, y2_pred))

6.911949297395278


In [165]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coefdf2= pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient'])


In [167]:
coefdf2.sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday       1.013816e+01
summit_elev        3.450061e+00
vertical_drop      2.531773e+00
clusters           2.347996e+00
daysOpenLastYear   2.130742e+00
TerrainParks       1.547432e+00
triple             1.399784e+00
surface            1.220272e+00
quad               9.978599e-01
SkiableTerrain_ac  9.422935e-01
projectedDaysOpen  8.436919e-01
Snow Making_ac     8.328663e-01
double             8.092155e-01
total_chairs       6.981532e-01
LongestRun_mi      6.933282e-01
averageSnowfall    4.361968e-01
NightSkiing_ac     3.643511e-01
Runs               3.502654e-01
yearsOpen          2.601287e-01
fastQuads          1.205103e-01
trams              1.332268e-15
fastSixes          8.881784e-16
fastEight          4.440892e-16

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [169]:
X3 = df2.drop(['Name','AdultWeekend', 'summit_elev'], axis=1)
y3 = df2.AdultWeekend
scaler3 = preprocessing.StandardScaler().fit(X3)
X_scaled3=scaler3.transform(X3) 

In [170]:
y3 = y3.ravel()
X3_train, X3_test, y3_train, y3_test = train_test_split(X_scaled3,y3, test_size=0.25, random_state=1)

In [171]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X3_train,y3_train)

In [172]:
y3_pred = model3.predict(X3_test)

In [173]:
print(y3_pred)

[79.69037882 55.56468145 59.50716642 68.02976703 51.6276937  48.10300798
 57.99031277 76.01284287 50.17358838 56.69144377 84.97330333 77.72614119
 58.42992534 74.33980766 31.05816771 57.24987973 47.73007137 50.22463406
 45.69649919 42.36390448 58.49885134 71.79333813 62.74058271 52.47098555
 41.59147965 40.16839208 53.56151068 55.8561322  79.40235818 39.54616635
 51.20020624 75.5070543  86.51097242 70.80099846 54.65891209 55.05725285
 32.47525465 69.63497751 51.78426712 41.42438429 32.76007938 43.82554346
 55.40895103 39.33055662]


In [174]:
print(model3.intercept_)

57.18946594035839


In [175]:
print(mean_absolute_error(y3_test, y3_pred))

6.486136118047323


In [176]:
print(explained_variance_score(y3_test, y3_pred))

0.768653881475073


In [177]:
coefdf3= pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient'])

In [178]:
coefdf3.sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday       1.004630e+01
vertical_drop      2.059360e+00
SkiableTerrain_ac  1.776930e+00
clusters           1.745455e+00
averageSnowfall    1.522437e+00
TerrainParks       1.373035e+00
daysOpenLastYear   1.356825e+00
Snow Making_ac     1.226439e+00
triple             1.195219e+00
surface            1.089996e+00
quad               9.346147e-01
LongestRun_mi      8.467776e-01
double             8.393082e-01
total_chairs       7.155975e-01
Runs               5.756870e-01
NightSkiing_ac     5.292723e-01
yearsOpen          4.868005e-01
fastQuads          3.113095e-01
projectedDaysOpen  2.788861e-02
fastEight          1.332268e-15
trams              8.881784e-16
fastSixes          8.881784e-16

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.750 | 6.64 |-|
| Model 2. | 0.754 | 6.91 |'state'|
| Model 3. | 0.769 | 6.49 |'state','summit_elev','base_elev'|

Model Selection: Model 3